### PR TITLE
Rebase #7223: improve contracts error message

### LINKED ISSUE #7209: [CT-2322] [Feature] Side-by-side diff when an enforced model contract fails
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing dbt functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

See this Loom starting at 12:50 for context of this feature proposal:
https://www.loom.com/share/f436e5e336e4433695dc63d53ee72fcf?t=770

Two things:
- Update "Contracts are enabled for this model." to "This model has an enforced contract."
- Do a side-by-side diff rather than a stacked listing of column names

Where a preflight check fails for a model contract, could the output look more like this?

image

Right now, the output looks like this:

image snip...
image

### Describe alternatives you've considered

Sometimes the status quo is the way to go ...
... and sometimes it isn't 😉 

### Who will this benefit?

This will benefit folks that:
- are using enforced model contracts
- there is a mismatch in the contract leading to a preflight failure
- have a model with tons of columns (or even just a few!)

### Are you interested in contributing this feature?

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by jtcohen6**
resolves #7209
rebases #7223

### Description

@kentkr did the heavy lifting here! I just added a few small touch-ups:
- Only display columns with diffs / mismatches
- Use `agate` for structured representation, and to print to stdout in a markdown table

```
09:52:16  Completed with 1 error and 0 warnings:
09:52:16
09:52:16  Compilation Error in model my_important_model (models/.../my_important_model.sql)
09:52:16    This model has an enforced contract that failed.
09:52:16    Please ensure the name, data_type, and number of columns in your contract match the columns in your model's definition.
09:52:16
09:52:16    | column_name                 | definition_type | contract_type | mismatch_reason       |
09:52:16    | --------------------------- | --------------- | ------------- | --------------------- |
09:52:16    | ADDRESS_ZIP                 | TEXT            | FIXED         | data type mismatch    |
09:52:16    | HAS_DOCS                    | FIXED           | BOOLEAN       | data type mismatch    |
09:52:16    | HAS_SUCCESSFUL_SESSION      | FIXED           |               | missing in contract   |
09:52:16    | HAS_VERY_SUCCESSFUL_SESSION |                 | FIXED         | missing in definition |
09:52:16
09:52:16
09:52:16    > in macro assert_columns_equivalent (macros/materializations/models/table/columns_spec_ddl.sql)
09:52:16    > called by macro default__get_assert_columns_equivalent (macros/materializations/models/table/columns_spec_ddl.sql)
09:52:16    > called by macro get_assert_columns_equivalent (macros/materializations/models/table/columns_spec_ddl.sql)
09:52:16    > called by macro snowflake__create_table_as (macros/adapters.sql)
09:52:16    > called by macro create_table_as (macros/materializations/models/table/create_table_as.sql)
09:52:16    > called by macro statement (macros/etc/statement.sql)
09:52:16    > called by macro materialization_table_snowflake (macros/materializations/table.sql)
09:52:16    > called by model my_important_model (models/.../my_important_model.sql)
```

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- ~I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR~
- [x] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

**Review by gshank — APPROVED** [review]

File: core/dbt/exceptions.py [hunk 1]
```diff
@@ -2145,14 +2147,65 @@ def __init__(self, yaml_columns, sql_columns):
...
+                    row = [sql_col["name"], sql_col["data_type"], "", "missing in contract"]
+                    mismatches += [dict(zip(column_names, row))]
+
+        # now add all yaml cols without a match
+        for yaml_col in self.yaml_columns:
+            if yaml_col["name"] not in sql_col_set:
+                row = [yaml_col["name"], "", yaml_col["data_type"], "missing in definition"]
+                mismatches += [dict(zip(column_names, row))]
+
+        mismatches_sorted = sorted(mismatches, key=lambda d: d["column_name"])
```
Comment (MichelleArk) [inline core/dbt/exceptions.py:2192]: 💯

**Review by MichelleArk — APPROVED** [review]

File: core/dbt/exceptions.py [hunk 2]
```diff
@@ -2145,14 +2147,65 @@ def __init__(self, yaml_columns, sql_columns):
...
+        # for each sql col list
+        for sql_col in self.sql_columns:
+            # add sql col to set
+            sql_col_set.add(sql_col["name"])
+            # for each yaml col list
+            for i, yaml_col in enumerate(self.yaml_columns):
+                # if name matches
+                if sql_col["name"] == yaml_col["name"]:
+                    # if type matches
+                    if sql_col["formatted"] == yaml_col["formatted"]:
```
Comment (MichelleArk) [inline core/dbt/exceptions.py:2168]: for my own understanding - any reason not to do the comparison on data_type at this point?

**Comment by kentkr** [comment]
Thanks @jtcohen6 !

**Comment by dbeatty10** [comment]
Nice work @kentkr ! 🏆

**Pull request merged**